core_backend/app/llm_call/parse_input.py

```diff
@@ -2,75 +2,196 @@
+        **kwargs: Any,
+    ) -> UserQueryResponse:
+        """
+        Wrapper function to identify the language of the question.
+        """
+        question, response = _identify_language(question, response)
+        response = await func(question, response, *args, **kwargs)
+        return response
+
+    return wrapper
+
+
+def _identify_language(
+    question: UserQueryRefined, response: UserQueryResponse
+) -> Tuple[UserQueryRefined, UserQueryResponse]:
     """
     Identifies the language of the question.
     """
+    if response.state != ResultState.ERROR:
+        question.original_language = getattr(
+            IdentifiedLanguage,
+            _ask_llm(question.query_text, IdentifiedLanguage.get_prompt()),
+        )
+        if question.original_language is not None:
+            response.debug_info["original_language"] = question.original_language.value
 
-    question.original_language = getattr(
-        IdentifiedLanguage,
-        _ask_llm(question.query_text, IdentifiedLanguage.get_prompt()),
-    )
+    return question, response
 
-    return question
 
+def translate_question(func: Callable) -> Callable:
+    """
+    Decorator to translate the question.
+    """
 
-def translate_question(question: UserQueryRefined) -> UserQueryRefined:
+    @wraps(func)
+    async def wrapper(
+        question: UserQueryRefined,
+        response: UserQueryResponse,
+        *args: Any,
+        **kwargs: Any,
+    ) -> UserQueryResponse:
+        """
+        Wrapper function to translate the question.
+        """
+        question, response = _identify_language(question, response)
+        question, response = _translate_question(question, response)
+        response = await func(question, response, *args, **kwargs)
+        response.debug_info["translated_question"] = question.query_text
+
+        return response
+
+    return wrapper
+
+
+def _translate_question(
+    question: UserQueryRefined, response: UserQueryResponse
+) -> Tuple[UserQueryRefined, UserQueryResponse]:
     """
     Translates the question to English.
     """
 
     if (
-        question.original_language
-        in [
-            IdentifiedLanguage.ENGLISH,
-            IdentifiedLanguage.UNKNOWN,
-        ]
-        or question.original_language is None
+        question.original_language == IdentifiedLanguage.ENGLISH
+        or response.state == ResultState.ERROR
     ):
-        return question
+        return question, response
+
+    if question.original_language is None:
+        response.state = ResultState.ERROR
+        raise ValueError(
+            (
+                "Language hasn't been identified. "
+                "Identify language before running translation"
+            )
+        )
+
+    if question.original_language == IdentifiedLanguage.UNKNOWN:
+        response.llm_response = (
+            "Sorry, we are unable to understand your question. "
+            "Please rephrase your question and try again."
+        )
```

Comment:
Should we be more informative about why we can't understand the question? Something like:

```suggestion
        response.llm_response = (
            f"Sorry, we are unable to understand your question. Currently, we can only understand {IdentifiedLanguage.keys()}"
        )
```
plus some string formatting and removing "unknown" from that list

Reply:
Sounds good!

Reply:
On a side note, "we" sounds like the Borg. I might change it to "this service" or "the bot".